overlord/ifacestate/hotplug.go

```diff
@@ -20,9 +20,47 @@
 package ifacestate
 
 import (
+	"strings"
+
 	"github.com/snapcore/snapd/interfaces/hotplug"
 )
 
+// List of attributes that determine the computation of default device key.
+// Attributes are grouped by similarity, the first non-empty attribute within the group goes into the key.
+// The final key is composed of 4 attributes (some of which may be empty), separated by "/".
+var attrGroups = [][]string{
+	// Name
+	{"ID_V4L_PRODUCT", "NAME", "ID_NET_NAME", "PCI_SLOT_NAME"},
+	// Vendor
+	{"ID_VENDOR_ID", "ID_VENDOR", "ID_WWN", "ID_WWN_WITH_EXTENSION", "ID_VENDOR_FROM_DATABASE", "ID_VENDOR_ENC", "ID_OUI_FROM_DATABASE"},
+	// Model
+	{"ID_MODEL_ID", "ID_MODEL_ENC"},
+	// Identifier
+	{"ID_SERIAL", "ID_SERIAL_SHORT", "ID_NET_NAME_MAC", "ID_REVISION"},
+}
+
+// defaultDeviceKey computes device key from the attributes of
+// HotplugDeviceInfo. Empty string is returned if too few attributes are present
+// to compute a good key. Attributes used to compute device key are defined in
+// attrGroups list above.
+func defaultDeviceKey(devinfo *hotplug.HotplugDeviceInfo) string {
```

Comment:
I agree, versioning makes sense. Also, +1 for using null instead of "/" as a separator, it could potentially lead to collisions indeed. I also changed the implementation to use sha checksums, the only downside of this that I see is it makes keys completely opaque, which makes debugging slightly harder.

One general and obvious issue I see is with reverts to old versions while new keys are in use in the state - we will not match affected devices.